x-pack/plugins/apm/server/lib/service_map/group_resource_nodes.ts

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { i18n } from '@kbn/i18n';
+import { groupBy } from 'lodash';
+import { ValuesType } from 'utility-types';
+import {
+  SPAN_TYPE,
+  SPAN_SUBTYPE,
+} from '../../../common/elasticsearch_fieldnames';
+import {
+  ConnectionNode,
+  isSpanGroupingSupported,
+} from '../../../common/service_map';
+
+const MINIMUM_GROUP_SIZE = 4;
+
+export function groupResourceNodes(responseData: {
+  elements: Array<{
+    data:
+      | ConnectionNode
+      | ({
```

Comment:
good idea, i might omit the `sourceData` and `targetData` properties from the edge definition types, since they're not actually used in the rendering.